ads/google/a4a/utils.js

```diff
@@ -0,0 +1,281 @@
+    new QueryParameter('eid', a4a.element.getAttribute('data-experiment-id')),
+    new QueryParameter('bih', viewportRect.height),
+    new QueryParameter('biw', viewportRect.width),
+    new QueryParameter('adx', slotRect.left),
+    new QueryParameter('ady', slotRect.top),
+    new QueryParameter('u_hist', getHistoryLength(global)),
+    dtdParam,
+    ...unboundedQueryParams,
+    new QueryParameter('url', documentInfo.canonicalUrl),
+    new QueryParameter(
+        'top', iframeDepth ? topWindowUrlOrDomain(global) : null),
+    new QueryParameter(
+        'loc',
+        global.location.href == documentInfo.canonicalUrl ?
+            null : global.location.href),
+    new QueryParameter('ref', referrer),
+  ];
+  dtdParam.value = elapsedTimeWithCeiling(timer.now(), startTime);
+  return buildUrl(
+      baseUrl, allQueryParams, MAX_URL_LENGTH,
+      new QueryParameter('trunc', '1'));
+}
+
+/**
+ * @param {!Window} win
+ * @return {!GoogleAdSlotCounter}
+ */
+export function getGoogleAdSlotCounter(win) {
+  if (!win.AMP_GOOGLE_AD_SLOT_COUNTER) {
+    win.AMP_GOOGLE_AD_SLOT_COUNTER = new GoogleAdSlotCounter();
+  }
+  return win.AMP_GOOGLE_AD_SLOT_COUNTER;
+}
+
+class GoogleAdSlotCounter {
+
+  constructor() {
+    /** @private {number} */
+    this.nextSlotNumber_ = 0;
+  }
+
+  /**
+   * @return {number}
+   */
+  nextSlotNumber() {
+    return ++this.nextSlotNumber_;
+  }
+
+}
+
+/**
+ * @param {!Window} global
+ * @return {number}
+ */
+function iframeNestingDepth(global) {
+  let win = global;
+  let depth = 0;
+  while (win != win.parent) {
+    win = win.parent;
+    depth++;
+  }
+  dev.assert(win == global.top);
+  return depth;
+}
+
+/**
+ * @param {number} slotNumber
+ * @param {!LayoutRectDef} slotRect
+ * @param {!LayoutRectDef} viewportRect
+ * @return {string}
+ */
+function adKey(slotNumber, slotRect, viewportRect) {
+  return formatFixedWidthInteger(slotNumber, 2) +
+    // ad slot top, in 1/5 viewport height units
+    formatFixedWidthInteger(slotRect.top * 5 / viewportRect.height, 4) +
+    // ad slot left, in 1/5 viewport width units
+    formatFixedWidthInteger(slotRect.left * 5 / viewportRect.width, 4);
+}
+
+/**
+ * @param {number} num Number, non-negative.
+ * @param {number} digits Number of digits, max 20.
+ * @return {string}
+ */
+function formatFixedWidthInteger(num, digits) {
+  const intPart = String(Math.max(Math.round(num), 0));
+  const len = intPart.length;
+  digits = Math.min(digits, 20);
+  if (len > digits) {
+    return '99999999999999999999'.substr(0, digits);
+  }
+  return '00000000000000000000'.substr(0, digits - len) + intPart;
+}
+
+/**
+ * @param {!Window} global
+ * @return {number}
+ */
+function getHistoryLength(global) {
+  // We have seen cases where accessing history length causes errors.
+  try {
+    return global.history.length;
+  } catch (e) {
+    return 0;
+  }
+}
+/**
+ * @param {!Window} global
+ * @return {?string}
+ */
+function topWindowUrlOrDomain(global) {
+  try {
```

Comment:
Please check first with `location.ancestorsOrigins` whether this can succeed by comparing against the current origin.